File: api/src/controllers/ingest.controller.ts
import { Request, Response } from 'express';
import { MinIOModel } from '../models/minio.model';
import { DocumentService } from '../services/document.service';
import { EmbeddingService } from '../services/embedding.service';
import { DatabaseModel } from '../models/database.model';

export class IngestController {
  constructor(
    private minio: MinIOModel,
    private documentService: DocumentService,
    private embeddingService: EmbeddingService,
    private db: DatabaseModel
  ) {}

  // 🚀 INGEST TOÀN BỘ BUCKET
  async ingestAll(req: Request, res: Response) {
    try {
      const basePrefix = 'courses-chatbot';
      const folders = ['curriculum', 'regulation', 'syllabus'];
       
      let totalFiles = 0;
      let totalChunks = 0;
      const errors: string[] = [];

      for (const folder of folders) {
        const prefix = `${basePrefix}/${folder}/`;
        
        console.log(`\n📂 Processing folder: ${prefix}`);
        
        try {
          const files = await this.minio.listFiles(prefix);
          console.log(`Found ${files.length} files in ${folder}`);

          for (const objectName of files) {
            // Bỏ qua thư mục
            if (objectName.endsWith('/')) continue;
            
            // Chỉ lấy file .docx
            if (!objectName.toLowerCase().endsWith('.docx')) {
              console.log(`⏭️ Skipping: ${objectName}`);
              continue;
            }

            console.log(`\n📥 Processing: ${objectName}`);
            
            try {
              totalFiles++;

              // 1. Download file
              const buffer = await this.minio.getFile(objectName);
              console.log(`  ✓ Downloaded ${buffer.length} bytes`);

              // 2. Extract text
              const rawText = await this.documentService.extractText(buffer, objectName);
              console.log(`  ✓ Extracted ${rawText.length} characters`);

              // 3. Clean text
              const text = this.documentService.cleanText(rawText);
              
              if (!text || text.trim().length === 0) {
                console.log(`  ⚠️ No content, skipping`);
                continue;
              }

              // 4. Parse metadata (FULL VERSION)
              const metadata = this.documentService.parseMetadataFromPath(
                objectName,
                text
              );

              console.log('  📋 Metadata:', JSON.stringify(metadata, null, 2));

              // 5. Insert document with full metadata
              const documentId = await this.db.insertDocument({
                filename: objectName.split('/').pop() || objectName,
                file_path: objectName,
                file_size: buffer.length,
                content_type: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
                document_type: metadata.document_type,
                metadata: metadata as any
              });

              console.log(`  ✓ Document inserted: ${documentId}`);

              // 6. Chunk text
              const chunks = this.documentService.chunkText(text, metadata);
              console.log(`  ✓ Created ${chunks.length} chunks`);

              // 7. Embed và save chunks
              for (let i = 0; i < chunks.length; i++) {
                const embedding = await this.embeddingService.generateEmbedding(chunks[i]);
                
                await this.db.insertChunk({
                  document_id: documentId,
                  content: chunks[i],
                  chunk_index: i,
                  embedding
                });

                totalChunks++;
                
                if ((i + 1) % 10 === 0) {
                  console.log(`  ✓ Embedded ${i + 1}/${chunks.length} chunks`);
                }
              }

              console.log(`✅ Successfully ingested: ${objectName}`);

            } catch (fileError: any) {
              const errorMsg = `Error processing ${objectName}: ${fileError.message}`;
              console.error(`❌ ${errorMsg}`);
              errors.push(errorMsg);
            }
          }
        } catch (folderError: any) {
          const errorMsg = `Error scanning folder ${folder}: ${folderError.message}`;
          console.error(`❌ ${errorMsg}`);
          errors.push(errorMsg);
        }
      }

      console.log(`\n🎉 Ingest completed!`);
      console.log(`📊 Total files: ${totalFiles}`);
      console.log(`📦 Total chunks: ${totalChunks}`);
      if (errors.length > 0) {
        console.log(`⚠️ Errors: ${errors.length}`);
      }

      res.json({
        message: 'Ingest hoàn tất',
        totalFiles,
        totalChunks,
        errors: errors.length > 0 ? errors : undefined,
        folders
      });

    } catch (error: any) {
      console.error('❌ Ingest failed:', error);
      res.status(500).json({ 
        error: error.message,
        stack: process.env.NODE_ENV === 'development' ? error.stack : undefined
      });
    }
  }

  // Kiểm tra trạng thái
  async checkIngestStatus(req: Request, res: Response) {
    try {
      const stats = await this.db.getIngestStats();
      res.json(stats);
    } catch (error: any) {
      console.error('Check status failed:', error);
      res.status(500).json({ error: error.message });
    }
  }

  // Xóa tất cả
  async clearAll(req: Request, res: Response) {
    try {
      const chunksResult = await this.db['pool'].query(
        'DELETE FROM chunks RETURNING id'
      );
      
      const docsResult = await this.db['pool'].query(
        'DELETE FROM documents RETURNING id'
      );
      
      console.log(`Deleted ${chunksResult.rowCount} chunks`);
      console.log(`Deleted ${docsResult.rowCount} documents`);
      
      res.json({
        message: 'Đã xóa toàn bộ dữ liệu',
        deletedChunks: chunksResult.rowCount,
        deletedDocuments: docsResult.rowCount
      });
    } catch (error: any) {
      console.error('Clear failed:', error);
      res.status(500).json({ error: error.message });
    }
  }
}

File: api/backend/controllers/ingest_controller.py
from __future__ import annotations

import os
from typing import Any

from backend.models.database import DatabaseModel
from backend.models.minio_model import MinIOModel
from backend.services.document_service import DocumentService
from backend.services.embedding_service import EmbeddingService


class IngestController:
    def __init__(
        self,
        *,
        minio: MinIOModel,
        document_service: DocumentService,
        embedding_service: EmbeddingService,
        db: DatabaseModel,
    ) -> None:
        self.minio = minio
        self.document_service = document_service
        self.embedding_service = embedding_service
        self.db = db

    def ingest_all(self, force: bool = False) -> tuple[dict[str, Any], int]:
        base_prefix = (os.getenv("MINIO_BASE_PREFIX") or "rag-processed").strip("/ ")
        ingest_batch_size = max(1, int(os.getenv("INGEST_BATCH_SIZE") or 32))
        folders = [
            "curriculum",
            "career",
            "career_description",
            "career-description",
            "career description",
            "regulation",
            "syllabus",
        ]
        supported_extensions = {".docx", ".pdf", ".txt", ".text", ".xlsx", ".xls"}

        candidate_files: set[str] = set()
        scanned_files = 0
        ingested_files = 0
        skipped_existing = 0
        total_chunks = 0
        errors: list[str] = []

        try:
            for folder in folders:
                prefix = f"{base_prefix}/{folder}/"
                try:
                    files = self.minio.list_files(prefix)
                except Exception as exc:
                    errors.append(f"Error scanning folder {folder}: {exc}")
                    continue

                for object_name in files:
                    scanned_files += 1
                    if object_name.endswith("/") or object_name.split("/")[-1].startswith("~$"):
                        continue
                    ext = f".{object_name.rsplit('.', 1)[-1].lower()}" if "." in object_name else ""
                    if ext not in supported_extensions:
                        continue
                    candidate_files.add(object_name)

            existing_paths: set[str] = set()
            if not force:
                try:
                    existing_paths = set(self.db.get_all_document_paths())
                except Exception as exc:
                    errors.append(f"Error loading existing document paths: {exc}")

            for object_name in sorted(candidate_files):
                if not force and object_name in existing_paths:
                    skipped_existing += 1
                    continue

                try:
                    buffer = self.minio.get_file(object_name)
                    raw_text = self.document_service.extract_text(buffer, object_name)
                    text = self.document_service.clean_text(raw_text)
                    if not text:
                        continue

                    metadata = self.document_service.parse_metadata_from_path(object_name, text)
                    document_id = self.db.insert_document(
                        filename=object_name.split("/")[-1],
                        file_path=object_name,
                        file_size=len(buffer),
                        content_type="application/octet-stream",
                        document_type=str(metadata.get("document_type") or "syllabus"),
                        metadata=metadata,
                    )

                    chunks = [chunk for chunk in self.document_service.chunk_text(text) if chunk]
                    for start in range(0, len(chunks), ingest_batch_size):
                        chunk_batch = chunks[start : start + ingest_batch_size]
                        embedding_batch = self.embedding_service.generate_batch_embeddings(chunk_batch)
                        self.db.insert_chunks_batch(
                            document_id=document_id,
                            chunks=chunk_batch,
                            embeddings=embedding_batch,
                            start_index=start,
                        )
                        total_chunks += len(chunk_batch)
                    ingested_files += 1
                except Exception as exc:
                    errors.append(f"Error processing {object_name}: {exc}")

            return {
                "message": "Ingest completed",
                "basePrefix": base_prefix,
                "scannedFiles": scanned_files,
                "candidateFiles": len(candidate_files),
                "ingestedFiles": ingested_files,
                "skippedExisting": skipped_existing,
                "totalChunks": total_chunks,
                "errors": errors if errors else None,
                "folders": folders,
                "force": force,
            }, 200
        except Exception as exc:
            return {"error": str(exc)}, 500

    def check_ingest_status(self) -> tuple[dict[str, Any], int]:
        try:
            return self.db.get_ingest_stats(), 200
        except Exception as exc:
            return {"error": str(exc)}, 500

    def clear_all(self) -> tuple[dict[str, Any], int]:
        try:
            self.db.clear_all()
            return {
                "message": "All indexed data cleared",
                "deletedChunks": None,
                "deletedDocuments": None,
            }, 200
        except Exception as exc:
            return {"error": str(exc)}, 500
